### PR TITLE
[WIP][lgwebos] Detection of power on/off with UPnP

### LINKED ISSUE #7119: [lgwebos] Enhancing the modelName property
A really minor enhancement would be to concatene "Model Name" and "Model Number" found by UPnP to set the thing property "modelName". Only "Model Name" is used until now which leads for my TV to this value "LG Smart TV" which is not really a model.

I will propose a fix.

### PR DESCRIPTION
**Pull request by lolodomo**
Also fix #7119

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
@sprehn : I let you check but it works very well for me.

I tested fully the following cases:

- thing created from the discovery in Paper UI (deviceId included in the discovery)
- thing created in Paper UI without the deviceId => thing configuration is then updated
- thing created from a config file with the deviceId setting set

In these cases, I see the Power channel immediately updated when TV switched on or off.

**Comment by sprehn** [comment]
on my TV this version does not give any improvement yet.

the following method gets called with status false exactly when the websocket connection breakes, 20-30seconds after the screen goes dark.

```
 public void onStatusChanged(boolean status) {
        logger.debug("onStatusChanged status={}", status);
        postUpdate(CHANNEL_POWER, status ? OnOffType.ON : OnOffType.OFF);
    }
```

in thingDiscovered I think we should check for the MediaRenderer with same host as our TV.
not THING_TYPE_WEBOSTV, cause that will disappear in UPNP too late.
In that case update the thing properties should not happen, as it will be a different device.
if we use the udn of media renderer then it should probably not be stored in deviceid config. it could either be a property, or even a variable in the handler.

**Comment by lolodomo** [comment]
Do you have the UPnP Device Spy app on your PC ?
I would like to be sure that with your TV, the UPnP devices are disappearing not all at the same time.
In case the MediaRenderer device is disappearing earlier and fast, we can decide to rely on this one rather than the other, that is not a big problem.

As I told you, in my case, I see the 4 UPnP devices appearing/disappearing immediately when I use the standby button of the remote control. So it does not matter what UPnP devices is considered by the discovery service.

In my case, the 4 UPnP devices have the same "friendly name". Note that I think we should rather use this "friendly name" to set the model name property.

As soon as you confirm that the MediaRenderer UPnP device is the one to consider because it appears/disappears very fast, I update the discovery code. This is only a minor change to do in the class LGWebOSUpnpDiscoveryParticipant. All the other code I introduced will remain unchanged.

**Comment by sprehn** [comment]
Hi,

using a macbook and Upnp Analyser, but this does not seem to capture those events nicely so that it actually helps me. To debug the situation I use the shutdown detector code:

```
@Override
    public void remoteDeviceRemoved(@Nullable Registry registry, @Nullable RemoteDevice device) {
        if (device == null) {
            return;
        }
        String ip = device.getIdentity().getDescriptorURL().getHost();
        thingRegistry.getAll().stream().filter(thing -> THING_TYPE_WEBOSTV.equals(thing.getThingTypeUID()))
                .filter(thing -> ip.equals(thing.getConfiguration().get(CONFIG_HOST))).forEach(thing -> {
                    ThingHandler handler = thing.getHandler();
                    if (handler != null) {
                        logger.debug("Detected device shutdown: {}", device);
                        logger.debug("device type: {}", device.getType());
                        for (RemoteService s : device.getServices()) {
                            logger.debug("service id: {}", s.getServiceId());
                            logger.debug("service type: {}", s.getServiceType());
                        }
                        ((LGWebOSHandler) handler).postUpdate(CHANNEL_POWER, OnOffType.OFF);
                    }
                });
    }
```

I see device urn:schemas-upnp-org:device:MediaRenderer:1 and its services immediatly and only 24 seconds later the service urn:lge-com:service:webos-second-screen:1 go down.

```2020-03-08 21:58:40.384 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - device type: urn:schemas-upnp-org:device:MediaRenderer:1
2020-03-08 21:58:40.384 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service id: urn:upnp-org:serviceId:AVTransport
2020-03-08 21:58:40.384 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service type: urn:schemas-upnp-org:service:AVTransport:1
2020-03-08 21:58:40.385 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service id: urn:upnp-org:serviceId:ConnectionManager
2020-03-08 21:58:40.385 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service type: urn:schemas-upnp-org:service:ConnectionManager:1
2020-03-08 21:58:40.385 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service id: urn:upnp-org:serviceId:RenderingControl
2020-03-08 21:58:40.385 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service type: urn:schemas-upnp-org:service:RenderingControl:1
2020-03-08 21:59:04.359 [DEBUG] [very.LGWebOSUpnpDiscoveryParticipant] - Found LG WebOS TV: (RemoteDevice) Identity: (RemoteDeviceIdentity) UDN: uuid:6da62662-1369-a82d-95c7-353c235ed503, Descriptor: http://192.168.2.119:1523/, Root: true
2020-03-08 21:59:04.359 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - Detected device shutdown: (RemoteDevice) Identity: (RemoteDeviceIdentity) UDN: uuid:6da62662-1369-a82d-95c7-353c235ed503, Descriptor: http://192.168.2.119:1523/, Root: true
2020-03-08 21:59:04.359 [DEBUG] [ebos.internal.handler.LGWebOSHandler] - Received thingRemoved event: lgwebos:WebOSTV:6da62662-1369-a82d-95c7-353c235ed503
2020-03-08 21:59:04.359 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - device type: urn:schemas-upnp-org:device:Basic:1
2020-03-08 21:59:04.360 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service id: urn:lge-com:serviceId:webos-second-screen-3000-3001
2020-03-08 21:59:04.360 [DEBUG] [iscovery.LGWebOSUpnpShutdownDetector] - service type: urn:lge-com:service:webos-second-screen:1
2020-03-08 21:59:04.397 [DEBUG] [bos.internal.handler.LGWebOSTVSocket] - WebSocket Closed - Code: 1001, Reason: server shutting down
```

If we changed what we listen for in LGWebOSUpnpDiscoveryParticipant, then it would change the deviceId for all current users, and we know that listening for urn:lge-com:service:webos-second-screen:1 works well for discovery. 
Users won't like such a change, as they will have to relink all channels to a new Thing.

btw. thingRemoved already got called without registering as UpnpIOParticipant or my ShutdownDetector when webos-second-screen disappears - only too late.

I thought the whole point was to use UpnpIOParticipant to react on a different device. 

But question would remain how we can discover it's UDN.

**Comment by sprehn** [comment]
why is implementing UpnpIOParticipant better than

**Comment by sprehn** [comment]
Would it be ok to go through all devices in the registry once connected and attempt to find the one which matches MediaRenderer on the same host:

```
this.upnpService.getRegistry().getDevices()
```

it would be sufficient to store that device's UDN in a property and we could then register the UpnpIOParticipant based on it.

**Comment by lolodomo** [comment]
If we change the UPnP device you are relying on in the discovery service, this will be fully transparent for any user and they will not have to relink channels. The deviceId is only a property of the current thing. It is also used as in the thing UID as default thing id but any user could have changed it.

I will update the code this evening, so that at least you can then confirm that it is working better for you.

Addtionnaly I will think if the discovery service can rely on the two UPnP devices.

**Comment by lolodomo** [comment]
@sprehn : can you please make a new try after my last change ?
Note that it is possible that you need to clear your inbox using the console command `smarthome:inbox clear`

It is possible that implementing UpnpIOParticipant is finally useless and that implementing DiscoveryListener could be sufficient.
First, I am waiting for your positive feedback and then I will simplify my PR.

**Comment by lolodomo** [comment]
Additionally I will evaluate if we can keep a simple property to store the UDN, It would avoid the edition of the thing configuration. I have the feeling it could work.

**Comment by sprehn** [comment]
Hi,
have to pause the work for the moment, as I don't have access to my TV to run tests.
This is due to day-care closure in Germany. Have moved temporary ...

**Comment by lolodomo** [comment]
I rebased my PR to have in my branch all the recently merged changes.

**Comment by lolodomo** [comment]
@sprehn : I have now terribly simplified my proposal. Everything is based on the discovery service. No need for additional JUPnP stuff, no need to transform the deviceId property into a configuration setting.
For me, it is working perfectly. I will wait for your feedback.
To summarize, the discovery is now based on the MediaRender device and the thing handler is considering the changes in the thing discovery to update the power channel.

**Comment by lolodomo** [comment]
Here ius a jar for testing:
[org.openhab.binding.lgwebos-2.5.3-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4334133/org.openhab.binding.lgwebos-2.5.3-SNAPSHOT.zip)

**Comment by lolodomo** [comment]
I rebased again to have a test version that includes all previous PR merged.
[org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4362632/org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip)

**Comment by lolodomo** [comment]
Here ius a jar for testing, it includes the fix to ignore a power off command when the TV is already off:
[org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4396824/org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip)

**Comment by lolodomo** [comment]
@sprehn : I integrated two fix:

1. update the key in the thing configuration only if the key is new. This avoids the thing update when the user has already set the right key.
2. abort the init process in case the thing handler is disposed while the init thread is still sending and receiving messages. This is done in the method handling the response. 

Here is a new jar for testing:
[org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4397423/org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip)

**Comment by sprehn** [comment]
Hi @lolodomo 

Re "Ignore power off command when the TV is already off"

So it seams the TV actually starts up again, if it receives a second "ssap://system/turnOff" message, during a specific time window while shutting down. Your proposed fix should prevent that we sent "ssap://system/turnOff"  twice, if the user sends two OFF events.

Now I wonder, will this be the same, if the user sends any another event, e.g. "ssap://tv/channelUp" while shutting down? If so, we should handle this further down in the socket implementation's state. I actually had a state DISCONNECTING at some point, which I later removed. We could use this to ignore any events being sent to the TV during shutdown.

**Comment by lolodomo** [comment]
@sprehn : will you review and validate at least few of my changes if I cut this PR in different PRs, even if you have no way to test ?

**Comment by lolodomo** [comment]
> Now I wonder, will this be the same, if the user sends any another event, e.g. "ssap://tv/channelUp" while shutting down? If so, we should handle this further down in the socket implementation's state. I actually had a state DISCONNECTING at some point, which I later removed. We could use this to ignore any events being sent to the TV during shutdown.

I updated my rule that was triggering the issue. I now increase the volume, mute it and set a new channel. I will see what happens after few days but I guess these commands will be just ignored and will not turn on the TV. like "ssap://system/turnOff" can do.

**Comment by lolodomo** [comment]
@sprehn : I created separate PRs with the hope we can merge few fixes quickly without your testing, while the main part of this PR will require your testing.

**Comment by lolodomo** [comment]
New rebase done to fix the conflict.

**Comment by lolodomo** [comment]
> > Now I wonder, will this be the same, if the user sends any another event, e.g. "ssap://tv/channelUp" while shutting down? If so, we should handle this further down in the socket implementation's state. I actually had a state DISCONNECTING at some point, which I later removed. We could use this to ignore any events being sent to the TV during shutdown.
> 
> I updated my rule that was triggering the issue. I now increase the volume, mute it and set a new channel. I will see what happens after few days but I guess these commands will be just ignored and will not turn on the TV. like "ssap://system/turnOff" can do.

@sprehn : I can confirm that I did not encounter any problem. I thing the problem is really solved now and was only a problem with the "ssap://system/turnOff" command.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java [hunk 1]
```diff
@@ -242,7 +259,7 @@ public String getKey() {
 
     @Override
     public void storeKey(@Nullable String key) {
-        if (!getKey().equals(key)) {
+        if (!disposed && !getKey().equals(key)) {
```
Comment (sprehn) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:262]: does this ever happen?

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java [hunk 2]
```diff
@@ -256,6 +273,9 @@ public void storeKey(@Nullable String key) {
 
     @Override
     public void storeProperties(Map<String, String> properties) {
+        if (disposed) {
```
Comment (sprehn) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:276]: does this ever happen?

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java [hunk 3]
```diff
@@ -55,18 +59,26 @@
...
-        if (device.findService(UPNP_SERVICE_TYPE) != null) {
+        // if (device.findService(UPNP_SERVICE_TYPE) != null) {
+        if (UPNP_MEDIARENDERER_DEVICE_TYPE.equals(device.getType())
+                && device.getDetails().getManufacturerDetails().getManufacturer() != null
+                && device.getDetails().getManufacturerDetails().getManufacturer().toUpperCase()
+                        .contains("LG ELECTRONICS")
+                && device.getDetails().getModelDetails().getModelDescription() != null
+                && device.getDetails().getModelDetails().getModelDescription().toUpperCase().contains("WEBOSTV")) {
             logger.debug("Found LG WebOS TV: {}", device);
             return new ThingUID(THING_TYPE_WEBOSTV, device.getIdentity().getUdn().getIdentifierString());
```
[the remark this answers is not in the document]
Comment (sprehn) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java:80]: I'll need to test this

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java [hunk 4]
```diff
@@ -147,13 +161,16 @@ public void dispose() {
...
+
         LGWebOSTVSocket s = socket;
         if (s != null) {
             s.setListener(null);
-            scheduler.execute(() -> s.disconnect()); // dispose should be none-blocking
+            s.disconnect();
         }
         socket = null;
         config = null; // ensure config gets actually refreshed during re-initialization
+        disposed = true;
```
Comment (sprehn) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:173]: why is this necessary?

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java [hunk 5]
```diff
@@ -328,4 +361,75 @@ private void refreshChannelSubscription(ChannelUID channelUID) {
...
+
+    @Override
+    public void thingDiscovered(DiscoveryService source, DiscoveryResult result) {
+        if (THING_TYPE_WEBOSTV.equals(result.getThingTypeUID())) {
+            Map<String, Object> properties = result.getProperties();
+            String host = (String) properties.get(CONFIG_HOST);
+            if (host != null && host.equals(getLGWebOSConfig().getHost())) {
+                // Thing matching the discovery
+
+                // Update the thing properties from the discovery result
```
[the remark this answers is not in the document]
Comment (sprehn) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:371]: so the media renderer will not have those additional properties I assume... ok, fine for me.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java [hunk 6]
```diff
@@ -334,4 +355,52 @@ public void postUpdate(String channelId, State state) {
...
+    public void thingDiscovered(DiscoveryService source, DiscoveryResult result) {
+        if (THING_TYPE_WEBOSTV.equals(result.getThingTypeUID())) {
+            Map<String, Object> properties = result.getProperties();
+            String host = (String) properties.get(CONFIG_HOST);
+            if (host != null && host.equals(getLGWebOSConfig().getHost())) {
+                // Thing matching the discovery
+                logger.debug("thingDiscovered: discovery with UID {} matching thing {}", result.getThingUID(),
+                        getThing().getUID());
+
+                channelHandlers.get(CHANNEL_POWER).onDeviceReady(CHANNEL_POWER, this);
```
Comment (sprehn) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:369]: websocket connected iff power on

for users, which don't use upnp this wont work.
it may create a situation, when power channel is on, but we are not connected yet.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java [hunk 7]
```diff
@@ -334,4 +355,52 @@ public void postUpdate(String channelId, State state) {
...
+                storeProperties(map);
+            }
+        }
+    }
+
+    @Override
+    public void thingRemoved(DiscoveryService source, ThingUID thingUID) {
+        if (!udn.isEmpty() && thingUID.getId().equals(udn)) {
+            logger.debug("thingRemoved: discovery with UID {} matching thing {}", thingUID, getThing().getUID());
+            channelHandlers.get(CHANNEL_POWER).onDeviceRemoved(CHANNEL_POWER, this);
```
Comment (sprehn) [inline bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:396]: same comment as in thing discovered.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 4 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:173]: This is the code from the PR #7299 
I integrated the other PR in this one for practical reason (for testing with all my changes) but this will be removed when this PR will be ready for a merge.
So please consider moving your comment in the right PR.

This is necessary because the websocket thread is closed asynchronously and so can call the thing handler (ConfigProvider) while the thing handler is already disposed.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 1 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:262]: This is the code from the PR #7299 So please consider moving your comment in the right PR.

This can happen because the websocket thread is closed asynchronously and so can call the thing handler (ConfigProvider) while the thing handler is already disposed.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:276]: This is the code from the PR #7299 So please consider moving your comment in the right PR.

This can happen because the websocket thread is closed asynchronously and so can call the thing handler (ConfigProvider) while the thing handler is already disposed.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSConfiguration.java [hunk 8]
```diff
@@ -28,6 +28,8 @@
     int port = 3000; // 3001 for TLS
     @Nullable
     String key; // name has to match LGWebOSBindingConstants.CONFIG_KEY
+    @Nullable
+    String deviceId; // name has to match LGWebOSBindingConstants.CONFIG_DEVICE_ID
```
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSConfiguration.java]: Apparently you commented an outdated version of the PR. deviceId is no more in the configuration object in the current PR.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:371]: If your thing is declared in a config file, it will have no properties set after startup. And it will not be set by the thing discovery.
Of course, the framework core code will check first if there are any change.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java:80]: I am not fully sure whether the previous discovery will be kept. Maybe just replaced.

But you certainly understood that the users will have nothing to change on their things. Their deviceId property will be updated automatically by the binding.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:371]: And of course, we absolutely need to change (if necessary) the deviceId property.
This allows an automatic update of all existing things when you move from the old code to the new code.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java [hunk 9]
```diff
@@ -58,15 +62,24 @@
                 .withProperty(PROPERTY_DEVICE_ID, device.getIdentity().getUdn().getIdentifierString())
                 .withProperty(CONFIG_HOST, device.getIdentity().getDescriptorURL().getHost())
                 .withLabel(device.getDetails().getFriendlyName())
-                .withProperty(PROPERTY_MODEL_NAME, device.getDetails().getModelDetails().getModelName())
+                // .withProperty(PROPERTY_MODEL_NAME,
+                // device.getDetails().getModelDetails().getModelName() + " "
+                // + device.getDetails().getModelDetails().getModelNumber())
+                .withProperty(PROPERTY_MODEL_NAME, device.getDetails().getFriendlyName())
```
[the remark this answers is not in the document]
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpDiscoveryParticipant.java:65]: Because they are not what we expect on this device.
For example, in my case:

- modelname = LG TV
- modelNumber = 1.0

friendlyName remains the same on all devices.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:396]: This line is the most important of this PR. It does what you expect since a long time, that is updating the power channel to OFF as fast as possible, even when the websocket is still connected.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 6 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:369]: I don't understand this part of your comment "for users, which don't use upnp this wont work.". The binding uses UPnP discovery, this is not a user's choice.
And for the few users having UPnP not working in their network, of course this PR will not help them as it is based on UPnP. But it probably concerns only a very small minority.

You are right, it can happen, at startup, that the channel is set to ON just before the websocket is connected and TV is considered as registered, as this is done in parallel by 2 different threads. 

But this is required to have a fast update of the channel when you power on the TV while the websocket was still alive and connected.
In fact, the goal of this PR is a fast update of the power channel (ON or OFF), not based on websocket connection status because the websocket remains connected during a certain time after the TV is switched to OFF.

But I agree that this channel represents whether the TV is ON or OFF, not whether the websocket is connected and TV is registered. This is an essential point.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 1 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java]: Removed from this PR

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java]: Removed from this PR

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 4 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java]: Removed from this PR

**Comment by lolodomo** [comment]
New rebase.
Here is a new jar for testing:
[org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip](https://github.com/openhab/openhab-addons/files/4463083/org.openhab.binding.lgwebos-2.5.4-SNAPSHOT.zip)

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 6 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:365]: Considering the PR #7355, I think that state update to ON will become unnecessary.

File: bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/handler/LGWebOSHandler.java:392]: Considering the PR #7355, I think we should trigger the new state DISCONNECTING in this case.

**Comment by lolodomo** [comment]
In fact, with PR #7355 handling the update of the power channel state, I think this PR could become simply useless.

**Comment by lolodomo** [comment]
The only interest will be the update of few properties we get only through UPnP discovery.

**Comment by lolodomo** [comment]
The main goal of this PR has disappeared after the merge of the PR #7355 
PR no more necessary.
